tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,1880 @@
+
+	/* Cloud is initialized */
+	init_cloud_success();
+
+	ret = nrf_cloud_sensor_data_send(&sensor_param_t);
+	zassert_equal(-EACCES, ret, "return should be -EACCES when not in dc_connected state");
+
+	/* Cloud is connected */
+	connect_cloud_success();
+
+	ret = nrf_cloud_sensor_data_send(&sensor_param_t);
+	zassert_equal(-EACCES, ret, "return should be -EACCES when not in dc_connected state");
+}
+
+/* Verify that the function will fail if a NULL param is provided */
+ZTEST(nrf_cloud_test, test_cloud_sensor_data_send_null_param)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully with dc_connect */
+	connect_cloud_dc_success();
+
+	int ret = nrf_cloud_sensor_data_send(NULL);
+
+	zassert_equal(-EINVAL, ret, "return should be -EINVAL with NULL param");
+}
+
+/* Verify that the function will fail if the
+ * cloud sensor data encoding function fails
+ */
+ZTEST(nrf_cloud_test, test_cloud_sensor_data_send_sensor_encode_failed)
+{
+	/* Sensor data for sending to cloud */
+	struct nrf_cloud_data cloud_data_t = {
+		.len = 0U,
+		.ptr = NULL,
+	};
+
+	struct nrf_cloud_sensor_data sensor_param_t = {
+		.type = NRF_CLOUD_SENSOR_GPS,
+		.data = cloud_data_t,
+		.tag = 20000U,
+	};
+
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully with dc_connect */
+	connect_cloud_dc_success();
+
+	nrf_cloud_encode_sensor_data_fake.custom_fake = fake_nrf_cloud_encode_sensor_data__fails;
+
+	int ret = nrf_cloud_sensor_data_send(&sensor_param_t);
+
+	zassert_not_equal(0, ret, "return should not be 0 when sensor data encoding fails");
+}
+
+/* Verify that the function will fail if the dc_send function fails */
+ZTEST(nrf_cloud_test, test_cloud_sensor_data_send_dc_send_failed)
+{
+	/* Sensor data for sending to cloud */
+	struct nrf_cloud_data cloud_data_t = {
+		.len = 0U,
+		.ptr = NULL,
+	};
+
+	struct nrf_cloud_sensor_data sensor_param_t = {
+		.type = NRF_CLOUD_SENSOR_GPS,
+		.data = cloud_data_t,
+		.tag = 20000U,
+	};
+
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	/* Init the cloud successfully */
+	init_cloud_success();
+
+	/* Connect to the cloud successfully with dc_connect */
+	connect_cloud_dc_success();
+
+	nrf_cloud_encode_sensor_data_fake.custom_fake = fake_nrf_cloud_encode_sensor_data__succeeds;
+	nct_dc_send_fake.custom_fake = fake_nct_dc_send__fails;
+
+	int ret = nrf_cloud_sensor_data_send(&sensor_param_t);
+
+	zassert_not_equal(0, ret, "return should not be 0 when sensor data sending fails");
+}
```

Comment:
This is also a test that should be in a separate suite that uses MQTT mocking instead of mocking the NCT/FSM, in my opinon